src/librustc_typeck/check/coercion.rs

```diff
@@ -60,28 +60,44 @@
 //! sort of a minor point so I've opted to leave it for later---after all
 //! we may want to adjust precisely when coercions occur.
 
-use check::{autoderef, FnCtxt, LvaluePreference, UnresolvedTypeAction};
+use check::{autoderef, impl_self_ty, FnCtxt, LvaluePreference, UnresolvedTypeAction};
 
 use middle::infer::{self, Coercion};
+use middle::subst::Substs;
 use middle::traits::{self, ObligationCause};
 use middle::traits::{predicate_for_trait_def, report_selection_error};
 use middle::ty::{AutoDerefRef, AdjustDerefRef};
 use middle::ty::{self, TypeAndMut, Ty, TypeError};
 use middle::ty_relate::RelateResult;
-use util::common::indent;
 
-use std::cell::RefCell;
 use std::collections::VecDeque;
 use syntax::ast;
 
 struct Coerce<'a, 'tcx: 'a> {
     fcx: &'a FnCtxt<'a, 'tcx>,
-    origin: infer::TypeOrigin,
-    unsizing_obligations: RefCell<Vec<traits::PredicateObligation<'tcx>>>,
+    origin: infer::TypeOrigin
 }
 
 type CoerceResult<'tcx> = RelateResult<'tcx, Option<ty::AutoAdjustment<'tcx>>>;
 
+/// The result of an attempt at coercing a Source type to a
+/// Target type via unsizing (`Source: CoerceUnsized<Target>`).
+/// If successful, all `CoerceUnsized` and `Unsized` obligations were selected.
+/// Other obligations, such as `T: Trait` for `&T -> &Trait`, are provided
+/// alongside the adjustment, to be enforced later.
+type CoerceUnsizedResult<'tcx> = Result<(AutoDerefRef<'tcx>,
```

Comment:
I believe I had tried that at some point and it got unwieldy, but maybe just a `struct` would work.